clang/docs/HIPSupport.rst

```diff
@@ -176,3 +176,95 @@ Predefined Macros
+
+Function Pointers Support
+=========================
+
+Function pointers' support varies with the usage mode in Clang with HIP. The following table provides an overview of the support status across different use-cases and modes.
+
+.. list-table:: Function Pointers Support Overview
+   :widths: 25 25 25
+   :header-rows: 1
+
+   * - Use Case
+     - ``-fno-gpu-rdc`` Mode (default)
+     - ``-fgpu-rdc`` Mode
+   * - Defined and used in the same TU
+     - Supported
+     - Supported
+   * - Defined in one TU and used in another TU
+     - Not Supported
+     - Supported
+
+In the ``-fno-gpu-rdc`` mode, the compiler calculates the resource usage of kernels based only on functions present within the same TU. This mode does not support the use of function pointers defined in a different TU due to the possibility of incorrect resource usage calculations, leading to undefined behavior.
+
+On the other hand, the ``-fgpu-rdc`` mode allows the definition and use of function pointers across different TUs, as resource usage calculations can accommodate functions from disparate TUs.
+
+Virtual Function Support
+========================
+
+In Clang with HIP, support for calling virtual functions of an object in device or host code is contingent on where the object is constructed.
+
+- **Constructed in Device Code**: Virtual functions of an object can be called in device code on a specific offloading device if the object is constructed in device code on an offloading device with the same architecture.
+- **Constructed in Host Code**: Virtual functions of an object can be called in host code if the object is constructed in host code.
+
+In other scenarios, calling virtual functions is not allowed.
+
+Explanation
+-----------
+
+An object constructed on the device side contains a pointer to the virtual function table on the device side, which is not accessible in host code, and vice versa. Thus, trying to invoke virtual functions from a context different from where the object was constructed will be disallowed because the appropriate virtual table cannot be accessed. The virtual function tables for offloading devices with different architecures are different, therefore trying to invoke virtual functions from an offloading device with a different architecture than where the object is constructed is also disallowed.
+
+A possible way to alleviate the current limitation of virtual function support in HIP is through the use of a "composite vtable". This involves creating a vtable that combines those from the host and all offloading device architectures, storing it in memory accessible by both. A dedicated registration function is introduced to populate this composite vtable. This function is invoked during global initialization to ensure the vtable is ready before any virtual function calls are made. For every virtual function call, irrespective of context, the system refers to this composite vtable to determine the correct function execution.
```

Comment:
Could you clarify whether the purpose of this pragraph is to provide a potential workaround to users or a solution in a future compiler?

Reply:
It is a possible solution. However, there may be better solutions that I am open to discussion.

Reply:
I would avoid speculating in this documentation.  Maybe we should remove this paragraph?

Reply:
removed the paragraph